waiter/src/waiter/auth/spnego.clj

```diff
@@ -83,32 +86,55 @@
    will be run, otherwise the handler will not be run and 401
    returned instead.  This middleware doesn't handle cookies for
    authentication, but that should be stacked before this handler."
-  [request-handler password]
-  (fn require-gss-handler [{:keys [headers] :as req}]
+  [request-handler ^ExecutorService thread-pool password]
+  (fn require-gss-handler [{:keys [headers] :as request}]
     (let [waiter-cookie (auth/get-auth-cookie-value (get headers "cookie"))
           [auth-principal _ :as decoded-auth-cookie] (auth/decode-auth-cookie waiter-cookie password)]
       (cond
         ;; Use the cookie, if not expired
         (auth/decoded-auth-valid? decoded-auth-cookie)
         (let [auth-params-map (auth/auth-params-map auth-principal)
               request-handler' (middleware/wrap-merge request-handler auth-params-map)]
-          (request-handler' req))
+          (request-handler' request))
         ;; Try and authenticate using kerberos and add cookie in response when valid
-        (get-in req [:headers "authorization"])
-        (let [^GSSContext gss_context (gss-context-init)
-              token (do-gss-auth-check gss_context req)]
-          (if (.isEstablished gss_context)
-            (let [principal (gss-get-princ gss_context)
-                  user (first (str/split principal #"@" 2))
-                  resp (auth/handle-request-auth request-handler req user principal password)]
-              (log/debug "added cookies to response")
-              (if token
-                (if (map? resp)
-                  (rr/header resp "WWW-Authenticate" token)
-                  (async/go
-                    (rr/header (async/<! resp) "WWW-Authenticate" token)))
-                resp))
-            (response-401-negotiate)))
+        (get-in request [:headers "authorization"])
+        (let [current-correlation-id (cid/get-correlation-id)
+              response-chan (async/promise-chan)
+              timer-context (timers/start (metrics/waiter-timer "core" "kerberos" "throttle" "delay"))]
+          ;; launch task that will populate the response in response-chan
+          (.execute
+            thread-pool
+            (fn process-gss-task []
+              (cid/with-correlation-id
+                current-correlation-id
+                (try
+                  (timers/stop timer-context)
+                  (let [^GSSContext gss-context (gss-context-init)
+                        token (do-gss-auth-check gss-context request)]
+                    (if (.isEstablished gss-context)
+                      (let [principal (gss-get-principal gss-context)
+                            user (first (str/split principal #"@" 2))
+                            response (auth/handle-request-auth request-handler request user principal password)]
```

Comment:
Processing the request should happen outside of the threadpool.  The async action should only cover authentication, returning only the principal through the channel, after which processing request can proceed.